aten/src/ATen/cuda/CUDAFuture.cpp

```diff
@@ -26,15 +26,26 @@ namespace {
 
 std::vector<std::reference_wrapper<const at::DataPtr>> extractDataPtrs(
     const at::IValue& value) {
-  at::IValue::HashAliasedIValues sub_values;
-  // Prefer getSubValues() over visit() as the latter is a silent no-op for
-  // some unsupported types, whereas the former at least fails loudly.
-  value.getSubValues(sub_values);
-
   std::vector<std::reference_wrapper<const at::DataPtr>> data_ptrs;
-  for (const at::IValue& sub_value : sub_values) {
-    if (sub_value.isTensor()) {
-      data_ptrs.emplace_back(sub_value.toTensor().storage().data_ptr());
+  // getSubValues works poorly on Python objects: it only works if they can be
+  // converted to a "regular" IValue type hence, for example, it doesn't support
+  // custom subclasses. Thus, instead, we extract the tensors through pickling.
+  if (value.isPyObject()) {
+    std::vector<at::Tensor> tensors =
+        value.toPyObjectHolder()->extractTensors();
```

Comment:
hmm, could you please remind me how does the build target dependency work? I thought this file is in libtorch_cuda and the `extractTensors` is in `libtorch_python`. If that's the case, does it mean we need to make CUDA target depends on Python target?

Reply:
So, I believe this works thanks to the fact that the `ConcretePyObjectHolder`, which is where the pickling actually happens, is in `libtorch_python` but its abstract base class `PyObjectHolder` is in `libtorch_cpp_cpu` (or whatever that's called). Thus CUDAFuture, which lives in `libtorch_cuda`, has access to the abstract base class, even though it doesn't have access to the concrete subclass. This is enough for the above code to build.